assignment5.cpp

```diff
@@ -0,0 +1,219 @@
+        if (a[index_a] != b[index_b]) {
+            first_not_common_chars.first = a[index_a];
+            first_not_common_chars.second = b[index_b];
+            break;
+        }
+        index_a++;
+        index_b++;
+    }
+    
+    return first_not_common_chars;
+}
+
+/* builds the adjacency matrix of the dag of chars */
+map<char, vector<char> > build_adjacency_matrix(vector<string> dictionary) {
+    map<char, vector<char> > adjacency_matrix;
+    
+    for (int i = 0; i < dictionary.size() - 1; i++) {
+        pair<char, char> first_not_common_chars = find_first_not_common_chars(dictionary[i], dictionary[i + 1]);
+        
+        if (first_not_common_chars.first != not_found) {
+            char first = first_not_common_chars.first;
+            char second = first_not_common_chars.second;
+            
+            if (!adjacency_matrix.count(first)) {
+                vector<char> adjacent_chars;
+                adjacent_chars.push_back(second);
+                adjacency_matrix.insert(make_pair(first, adjacent_chars));
+            } else {
+                adjacency_matrix[first].push_back(second);
+            }
+            
+        }
+        
+    }
+    
+    return adjacency_matrix;
+    
+}
+
+/* finds the in degrees of the nodes in the dag */
+map<char, int> find_in_degrees(map<char, vector<char> > adjacency_matrix) {
+    map<char, int> in_degrees;
+    
+    for (map<char, vector<char> >::iterator it = adjacency_matrix.begin(); it!=adjacency_matrix.end(); ++it) {
+        vector<char> adj_chars = it -> second;
+        for (int i = 0; i < adj_chars.size(); i++) {
+            if (!in_degrees.count(adj_chars[i])) {
+                in_degrees.insert(make_pair(adj_chars[i], 1));
+            } else {
+                in_degrees[adj_chars[i]]++;
+            }
+        }
+    }
+    
+    return in_degrees;
+}
+
+/* visits node in the dag */
+void visit(char char_to_visit, map<char, int> *in_degrees, set<char> *visited, map<char, vector<char> > adj_matrix, queue<char> *roots_of_dag) {
+    vector<char> adj_chars = adj_matrix[char_to_visit];
+    
+    visited->insert(char_to_visit);
+    
+    for (int i = 0; i < adj_chars.size(); i++) {
+        if (!visited->count(adj_chars[i])) {
+            (*in_degrees)[adj_chars[i]]--;
+            if ((*in_degrees)[adj_chars[i]] == 0) {
+                roots_of_dag->push(adj_chars[i]);
+            }
+        }
+    }
+}
+
+/* does the topological sorting and returns the ordered alphabet */
+vector<char> topological_sort(map<char, vector<char> > adjacency_matrix, map<char, int> in_degrees, vector<string> dictionary) {
+    vector<char> ordered_alphabet;
+    set<char> visited;
+    queue<char> roots_of_dag;
+    
+    /* adds all chars with 0 in degree to the queue */
+    for (int i = 0; i < dictionary.size(); i++) {
+        for (int j = 0; j < dictionary[i].size(); j++) {
+            if (!in_degrees.count(dictionary[i][j])) {
```

Comment:
You can iterate on in_degrees values instead of all the chars of all the letters (many duplications).
This way you won't add each letter many times to the queue:
```C++
for (pair<char, int> entry : in_degrees) {
  if (entry.second == 0) {  // The in_degree is 0.
    roots_of_dag.push(entry.second);  // Add the char to the roots queue.
  }
}
```
This way your queue won't have duplications.
e.g. for dictionary `{"aaa", "aab"}`, your queue will be just `{'a'}` instead of `{'a','a','a','a','a','a'}`.